docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+by non-Google contributors is low. OPA is planning to replace their policy
+language Rego with CEL, but
+[no timeline has been announced for this migration](https://github.com/open-policy-agent/opa/issues/811#issuecomment-401844999).
+The KrakenD API Gateway has announced support for CEL
+[here](https://medium.com/devops-faith/krakend-api-gateway-0-9-released-9427c249dbcd).
+
+The CEL language is new and has limited exposure, despite being informed by
+experience and research at Google. We may discover issues that make it less
+suitable for Trigger filtering (e.g.
+[google/cel-go#203](https://github.com/google/cel-go/issues/203).
+
+If CEL turns out to be a liability, a new expression language could be added
+alongside CEL without breaking existing triggers, allowing for an orderly
+deprecation.
+
+For example, a javascript filter could be added without disrupting existing CEL
+filters:
+
+```yaml
+spec:
+  filter:
+    javascript:
+      expression: ce.type == "com.github.pull.create"
+```
+
+## Alternatives Considered
+
+The following alternative solutions were considered but rejected.
+
+### Javascript or Lua as expression language
+
+Trigger filtering is a very limited case for an embedded language. All that is
+needed from a language is the ability to write conditions that evaluate to true
+or false.
+
+Javascript and Lua are commonly used as embedded languages, but their
+capabilities extend well beyond the needs of Trigger filtering. Those
+capabilities come with complexity: Javascript and Lua programs may not provably
+terminate, may have memory leaks, and may consume excessive amounts of CPU. All
+these cases would need to be guarded against by any process involved in
+filtering. If the expression language is used for upstream filter propagation,
+this could be a large number of diverse processes.
+
+Due to this additional implementation complexity, both Javascript and Lua have
+been judged less suitable for Trigger filtering than CEL.
+
+### Rego (OPA)
+
+[Rego](https://www.openpolicyagent.org/docs/v0.10.7/language-reference/) is
+already as a policy expression language by
+[Open Policy Agent](https://www.openpolicyagent.org/). It could be an attractive
+choice, but
+[OPA has indicated a desire to replace Rego with CEL](https://github.com/open-policy-agent/opa/issues/811#issuecomment-401844999).
+Since Rego doesn't seem to be used outside OPA, choosing Rego for Trigger
+filters would likely make Triggers the only remaining use of Rego.
+
+The possible maintenance burden of being the only Rego user makes Rego less
+suitable for Trigger filtering than CEL.
+
+### Custom expression language
+
+We could develop our own custom expression language for Trigger filters. This
+might be attractive since we could design the language for the needs of Trigger
+filters specifically.
+
+A possible example is Kubernetes' set-based label selectors:
+
+```
+environment in (production, qa)
+tier notin (frontend, backend)
+partition
+!partition
+```
```

Comment:
For the most part the above expressions can be implemented with CEL with very few modifications:

```
environment in ["production", "qa"]
!(tier in ["frontend", "backend"])
partition
!partition
```